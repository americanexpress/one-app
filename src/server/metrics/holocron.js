/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createCounter } from './counters';
import { createGauge } from './gauges';

import createMetricNamespace from './create-metric-namespace';

const holocronNamespace = createMetricNamespace('holocron');

createCounter({
  name: holocronNamespace('module_map_poll', 'total'),
  help: 'total times the module map has been polled',
});

createCounter({
  name: holocronNamespace('module_map_updated', 'total'),
  help: 'total times the module map has been updated',
});

createCounter({
  name: holocronNamespace('module_map_poll_restarted', 'total'),
  help: 'total times the module map poll has been restarted',
});

createGauge({
  name: holocronNamespace('module_map_poll_wait', 'seconds'),
  help: 'delay until the next poll of the module map',
});

createGauge({
  name: holocronNamespace('module_map_poll_consecutive_errors', 'total'),
  help: 'how many times has the server encountered errors polling the module map',
});

export default holocronNamespace.getMetricNames();
