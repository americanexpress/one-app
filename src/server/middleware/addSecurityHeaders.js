/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export default function addSecurityHeaders(req, res, next) {
  res.set('X-Frame-Options', 'DENY');
  res.set('X-Content-Type-Options', 'nosniff');
  res.set('Strict-Transport-Security', 'max-age=15552000; includeSubDomains');
  res.set('X-XSS-Protection', '1; mode=block');
  res.set('Referrer-Policy', process.env.ONE_REFERRER_POLICY_OVERRIDE || 'same-origin');
  next();
}
