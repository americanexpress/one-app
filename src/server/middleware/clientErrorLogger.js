/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// /_/report/errors
import util from 'util';

const nodeEnvIsDevelopment = process.env.NODE_ENV === 'development';

export default function clientErrorLogger(req, res) {
  if (!nodeEnvIsDevelopment) {
    const contentType = req.headers['content-type'];
    if (!/^application\/json/i.test(contentType)) {
      return res.sendStatus(415);
    }

    const {
      body: errorsReported,
      headers: {
        'correlation-id': correlationId,
        'user-agent': userAgent,
      },
    } = req;

    if (Array.isArray(errorsReported)) {
      errorsReported.forEach((raw) => {
        if (!raw || typeof raw !== 'object') {
          // drop on the floor, this is the wrong interface
          console.warn(`dropping an error report, wrong interface (${typeof raw})`);
          return;
        }
        const {
          msg, stack, href, otherData,
        } = raw;
        const err = new Error(msg);
        Object.assign(err, {
          name: 'ClientReportedError',
          stack,
          userAgent,
          uri: href,
          metaData: {
            ...otherData,
            correlationId,
          },
        });
        console.error(util.inspect(err, false, 10, true));
      });
    } else {
      // drop on the floor, this is the wrong interface
      console.warn(`dropping an error report group, wrong interface (${typeof errorsReported})`);
    }
  }

  return res.sendStatus(204);
}
