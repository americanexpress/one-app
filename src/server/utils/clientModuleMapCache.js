/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

let cache = {};

/* Filters bundle types so only the required client moduleBundleType is returned
in ../server/middleware/sendHtml.js and reducing the html payload size */
function filterBundles(moduleMap, moduleBundleType) {
  return {
    ...moduleMap,
    modules: Object.entries(moduleMap.modules).reduce((acc, [moduleName, moduleBundles]) => (
      {
        ...acc,
        [moduleName]: {
          baseUrl: moduleBundles[moduleBundleType].url.replace(/[^/]+\.js$/i, ''),
          [moduleBundleType]: moduleBundles[moduleBundleType],
        },
      }
    ), {}),
  };
}

export function setClientModuleMapCache(moduleMap) {
  cache = {
    browser: filterBundles(moduleMap, 'browser'),
    legacyBrowser: filterBundles(moduleMap, 'legacyBrowser'),
  };
}

export function getClientModuleMapCache() {
  return cache;
}
