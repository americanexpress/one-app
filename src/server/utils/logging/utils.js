/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
