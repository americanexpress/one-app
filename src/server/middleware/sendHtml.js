/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// This rule is only needed for a couple functions below
/* eslint-disable es/no-arrow-functions */
import { Set as iSet, Map as iMap } from 'immutable';

import transit from '../../universal/utils/transit';
import { setConfig } from '../../universal/ducks/config';
import jsonStringifyForScript from '../utils/jsonStringifyForScript';
import { getClientStateConfig } from '../utils/stateConfig';
import getI18nFileFromState from '../utils/getI18nFileFromState';
import renderModuleStyles from '../utils/renderModuleStyles';
import readJsonFile from '../utils/readJsonFile';
import { getClientPWAConfig } from './pwa/config';

const { buildVersion } = readJsonFile('../../../.build-meta.json');
const integrityManifest = readJsonFile('../../../bundle.integrity.manifest.json');
const nodeEnvIsDevelopment = process.env.NODE_ENV === 'development';

// http://www.useragentstring.com/pages/useragentstring.php?name=Internet+Explorer
const legacyIndicators = [
  'rv:11', // IE 11 on mobile
  'MSIE', // IE
];

const legacyUserAgent = (userAgent) => legacyIndicators
  .some((legacyIndicator) => userAgent.includes(legacyIndicator));
function getChunkAssets(assetsByChunkName) {
  return Object
    .entries(assetsByChunkName)
    // i18n is different per request, app needs to be the last chunk loaded
    .filter(([chunkName]) => !chunkName.startsWith('i18n/') && !['app', 'service-worker-client'].includes(chunkName))
    .map(([, assets]) => (typeof assets === 'string' ? assets : assets[0]));
}

const modernBrowserChunkAssets = getChunkAssets(readJsonFile('../../../.build-meta.json').modernBrowserChunkAssets);
const legacyBrowserChunkAssets = getChunkAssets(readJsonFile('../../../.build-meta.json').legacyBrowserChunkAssets)
  .map((chunkAsset) => `legacy/${chunkAsset}`);

export function safeSend(res, ...payload) {
  if (!res.headersSent) {
    res.send(...payload);
  }
}

export function renderStaticErrorPage(res) {
  if (!res.statusCode) {
    res.status(500);
  }
  console.info(`renderStaticErrorPage status ${res.statusCode}`);

  let message = 'Sorry, we are unable to load this page at this time. Please try again later.';
  if (res.statusCode >= 400 && res.statusCode < 500 && res.statusCode !== 404) {
    // issue is with the request, retrying won't change the server response
    message = 'Sorry, we are unable to load this page at this time.';
  }

  // TODO: allow root module to provide custom error message and override default html
  safeSend(res,
    `<!DOCTYPE html>
        <html>
          <head>
            <title>One App</title>
            <meta http-equiv="X-UA-Compatible" content="IE=edge">
            <meta charset="utf-8">
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <meta name="application-name" content="one-app">
          </head>
          <body style="background-color: #F0F0F0">
            <div id="root">
              <div>
                <div style="width: 70%; background-color: white; margin: 4% auto;">
                  <h2 style="display: flex; justify-content: center; padding: 40px 15px 0px;">Loading Error</h2>
                  <p style="display: flex; justify-content: center; padding: 10px 15px 40px;">
                    ${message}
                  </p>
                </div>
              </div>
            </div>
          </body>
        </html>`);
}

function renderI18nScript(clientInitialState, appBundlesURLPrefix) {
  const i18nFile = getI18nFileFromState(clientInitialState);
  if (!i18nFile) {
    return '';
  }

  return `<script src="${appBundlesURLPrefix}/${i18nFile}" crossorigin="anonymous"></script>`;
}

export function renderModuleScripts({
  clientInitialState, moduleMap, isDevelopmentEnv, bundle,
}) {
  const clientConfig = getClientStateConfig();
  const { rootModuleName } = clientConfig;
  // Sorting to ensure that the rootModule is the first script to load,
  // this is required to correctly provide external dependencies.
  const orderedLoadedModules = clientInitialState.getIn(['holocron', 'loaded'], iSet())
    .sort((currentModule, nextModule) => {
      if (currentModule === rootModuleName) { return -1; }
      if (nextModule === rootModuleName) { return 1; }
      return 0;
    });

  return orderedLoadedModules.map((moduleName) => {
    const { integrity, url: src } = moduleMap.modules[moduleName][bundle];
    const { clientCacheRevision } = moduleMap;
    const additionalAttributes = isDevelopmentEnv ? '' : `integrity="${integrity}"`;
    const scriptSource = isDevelopmentEnv || !clientCacheRevision ? src : `${src}?clientCacheRevision=${clientCacheRevision}`;
    return `<script src="${scriptSource}" crossorigin="anonymous" ${additionalAttributes}></script>`;
  }).join(isDevelopmentEnv ? '\n          ' : '');
}

function serializeClientInitialState(clientInitialState) {
  // try to build the full state, this _might_ fail (ex: 'Error serializing unrecognized object')
  try {
    return transit.toJSON(clientInitialState);
  } catch (err) {
    console.error('encountered an error serializing full client initial state', err);

    // clear out an internal cache that corrupts the serialization generated on the next call
    // TODO: understand transit-js and transit-immutable-js internals to properly fix the bug
    // for now stop the bleeding
    transit.toJSON('clear out an internal cache');
  }

  // can't send all the work we've done to build the state, but we can still give the app what it
  // needs to start in the browser
  // this _shouldn't_ ever throw, but just in case...
  try {
    return transit.toJSON(iMap({
      config: clientInitialState.get('config'),
      holocron: clientInitialState.get('holocron'),
    }));
  } catch (err) {
    transit.toJSON('clear out an internal cache, again');
    // something is really wrong
    console.error('unable to build the most basic initial state for a client to startup', err);
    throw err;
  }
}

function getHelmetData(data, fallback = '') {
  return (data && data.toString()) || fallback;
}

function getHelmetString(helmetInfo, disableStyles) {
  const {
    title, meta, style, link, base,
  } = helmetInfo;

  let linkText = getHelmetData(link);

  // filter only stylesheets
  if (linkText && disableStyles) {
    linkText = linkText.match(/<[^>]+>/g)
      .filter((match) => !match.includes('rel="stylesheet"'))
      .join('');
  }

  const styleText = disableStyles ? '' : getHelmetData(style);

  return `
${getHelmetData(title, '<title>One App</title>')}
${getHelmetData(meta)}
${styleText}
${linkText}
${getHelmetData(base)}
  `;
}

export function getHtmlAttributesString(helmetInfo) {
  const { htmlAttributes } = helmetInfo;
  return (htmlAttributes && ` ${htmlAttributes.toString()}`) || '';
}

export function getHead({
  helmetInfo,
  store,
  disableStyles,
  pwaMetadata,
}) {
  return `
    <head>
      ${getHelmetString(helmetInfo, disableStyles)}
      ${disableStyles ? '' : `
      ${renderModuleStyles(store)}
      `}
      ${pwaMetadata.webManifestUrl ? `<link rel="manifest" href="${pwaMetadata.webManifestUrl}">` : ''}
    </head>
  `;
}

export function getBody({
  isLegacy,
  helmetInfo,
  renderMode,
  assets,
  appHtml,
  appBundlesURLPrefix,
  clientInitialState,
  disableScripts,
  clientModuleMapCache,
  scriptNonce,
  pwaMetadata,
}) {
  const bundle = isLegacy ? 'legacyBrowser' : 'browser';
  const { bodyAttributes, script } = helmetInfo;
  const bundlePrefixForBrowser = isLegacy ? `${appBundlesURLPrefix}/legacy` : appBundlesURLPrefix;
  return `
    <body${(bodyAttributes && ` ${bodyAttributes.toString()}`) || ''}>
      <div id="root">${appHtml || ''}</div>
      ${disableScripts ? '' : `
      <script id="initial-state" ${scriptNonce ? `nonce="${scriptNonce}"` : ''}>
        window.__webpack_public_path__ = ${jsonStringifyForScript(`${appBundlesURLPrefix}/`)};
        window.__CLIENT_HOLOCRON_MODULE_MAP__ = ${jsonStringifyForScript(clientModuleMapCache[bundle])};
        window.__INITIAL_STATE__ = ${jsonStringifyForScript(serializeClientInitialState(clientInitialState))};
        window.__holocron_module_bundle_type__ = '${bundle}';
        window.__pwa_metadata__ = ${jsonStringifyForScript(pwaMetadata)};
        window.__render_mode__ = '${renderMode}';
      </script>
      ${assets}
      ${renderI18nScript(clientInitialState, bundlePrefixForBrowser)}
      ${renderModuleScripts({
    clientInitialState,
    moduleMap: clientModuleMapCache[bundle],
    isDevelopmentEnv: nodeEnvIsDevelopment,
    bundle,
  })}
      <script src="${bundlePrefixForBrowser}/app.js" integrity="${integrityManifest[isLegacy ? 'legacy/app.js' : 'app.js']}" crossorigin="anonymous"></script>
      ${(script && script.toString()) || ''}
      `}
    </body>
  `;
}

export function renderPartial({
  html: initialHtml,
  store,
}) {
  let html = initialHtml;
  // The partial route is commonly used for HTML partials, however email rendering requires
  // generating the entire document (including the <!doctype>), not just a portion of a document.
  // As Fragments cannot use dangerouslySetInnerHTML an outer element is required, but not desired.
  // In the final content body, we use `<dangerously-return-only-doctype>`.
  if (
    initialHtml.match(/^<dangerously-return-only-doctype>\s*<!doctype html>/)
    && initialHtml.match(/<\/html>\s*<\/dangerously-return-only-doctype>$/)
  ) {
    html = initialHtml
      .replace(/^<dangerously-return-only-doctype>/, '')
      .replace(/<\/dangerously-return-only-doctype>$/, '');
  }

  const styles = renderModuleStyles(store);

  if (html.startsWith('<!doctype html>')) {
    return html.replace('</head>', `${styles}</head>`);
  }

  return styles + html;
}

// TODO add additional client side scripts
export default function sendHtml(req, res) {
  let body;
  try {
    const {
      appHtml, clientModuleMapCache, store, headers, helmetInfo = {}, renderMode = 'hydrate',
    } = req;
    const { scriptNonce } = res;
    const userAgent = headers['user-agent'];
    const isLegacy = legacyUserAgent(userAgent);

    console.info(`sendHtml, have store? ${!!store}, have appHtml? ${!!appHtml}`);
    if (appHtml && typeof appHtml !== 'string') {
      throw new Error(`appHtml was not a string, was ${typeof appHtml}`, appHtml);
    }
    // replace server specific config with client specific config (api urls and such)
    const clientConfig = getClientStateConfig();
    const pwaMetadata = getClientPWAConfig();
    store.dispatch(setConfig(clientConfig));
    const cdnUrl = clientConfig.cdnUrl || '/_/static/';
    const clientInitialState = store.getState();
    const appBundlesURLPrefix = `${cdnUrl}app/${buildVersion}`;

    const disableScripts = clientInitialState.getIn(['rendering', 'disableScripts']);
    const disableStyles = clientInitialState.getIn(['rendering', 'disableStyles']);
    const renderPartialOnly = clientInitialState.getIn(['rendering', 'renderPartialOnly']);

    if (renderPartialOnly) {
      return safeSend(res, renderPartial({ html: req.appHtml, store }));
    }

    const chunkAssets = isLegacy ? legacyBrowserChunkAssets : modernBrowserChunkAssets;

    const assets = chunkAssets
      .map((chunkAsset) => `<script src="${appBundlesURLPrefix}/${chunkAsset}" integrity="${integrityManifest[chunkAsset]}" crossorigin="anonymous"></script>`)
      .join('\n          ');

    const headSectionArgs = {
      helmetInfo,
      store,
      disableScripts,
      disableStyles,
      scriptNonce,
      pwaMetadata,
    };

    const bodySectionArgs = {
      helmetInfo,
      renderMode,
      isLegacy,
      assets,
      appHtml,
      appBundlesURLPrefix,
      clientInitialState,
      disableScripts,
      clientModuleMapCache,
      scriptNonce,
      pwaMetadata,
    };

    body = `
      <!DOCTYPE html>
      <html${getHtmlAttributesString(helmetInfo)}>
        ${getHead(headSectionArgs)}
        ${getBody(bodySectionArgs)}
      </html>
    `;
  } catch (err) {
    console.error('sendHtml had an error, sending static error page', err);
    return renderStaticErrorPage(res);
  }

  return safeSend(res, body);
}
