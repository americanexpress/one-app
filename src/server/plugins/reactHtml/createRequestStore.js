/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createHolocronStore } from 'holocron';
import { fromJS, Set as iSet } from 'immutable';

import renderStaticErrorPage from './staticErrorPage';
import createEnhancer from '../../../universal/enhancers';
import safeRequest from '../../utils/safeRequest';
import { getServerStateConfig, getClientStateConfig } from '../../utils/stateConfig';
import createSsrFetch from '../../utils/createSsrFetch';
import { getClientModuleMapCache } from '../../utils/clientModuleMapCache';

const createRequestStore = (
  request, reply,
  { reducers }
) => {
  // fastify.decorateRequest('store', null);
  // fastify.decorateRequest('clientModuleMapCache', null);

  try {
    const serverConfig = getServerStateConfig();
    const clientConfig = getClientStateConfig();

    const initialServerState = fromJS({
      holocron: { loaded: iSet() },
      config: serverConfig,
    });

    const fetchClient = createSsrFetch({
      req: request.raw,
      res: reply.raw,
    })(fetch);

    console.log('--request.method', request.method);
    console.log('--request.body', request.body);
    console.log('--request.raw.body', request.raw.body);

    const enhancer = createEnhancer();
    const localsForBuildInitialState = {
      req: safeRequest(request, { useBodyForBuildingTheInitialState: request.method.toLowerCase() === 'post' }),
      config: clientConfig,
    };

    const store = createHolocronStore({
      reducer: reducers,
      initialState: initialServerState.merge(
        fromJS(reducers.buildInitialState(localsForBuildInitialState))
      ),
      enhancer,
      localsForBuildInitialState,
      extraThunkArguments: { fetchClient },
    });

    // TODO: namespace?
    // use the store as a global for the request
    request.store = store; // eslint-disable-line no-param-reassign
    request.clientModuleMapCache = getClientModuleMapCache();
  } catch (err) {
    console.error('error creating store for request', err);
    // TODO: migrate `renderStaticErrorPage`
    return renderStaticErrorPage(reply);
  }
};

export default createRequestStore;
