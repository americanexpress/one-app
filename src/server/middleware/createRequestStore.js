/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createHolocronStore } from 'holocron';
import { fromJS, Set as iSet } from 'immutable';

import { renderStaticErrorPage } from './sendHtml';
import createEnhancer from '../../universal/enhancers';
import safeRequest from '../utils/safeRequest';
import { getServerStateConfig, getClientStateConfig } from '../utils/stateConfig';
import createSsrFetch from '../utils/createSsrFetch';
import { getClientModuleMapCache } from '../utils/clientModuleMapCache';

export default function createRequestStore(
  { reducers },
  { useBodyForBuildingTheInitialState = false } = {}
) {
  return (req, res, next) => {
    try {
      const serverConfig = getServerStateConfig();
      const clientConfig = getClientStateConfig();

      const initialServerState = fromJS({
        holocron: { loaded: iSet() },
        config: serverConfig,
      });

      const fetchClient = createSsrFetch({
        req,
        res,
      })(fetch);

      const enhancer = createEnhancer();
      const localsForBuildInitialState = {
        req: safeRequest(req, { useBodyForBuildingTheInitialState }),
        config: clientConfig,
      };
      const store = createHolocronStore({
        reducer: reducers,
        initialState: initialServerState.merge(
          fromJS(reducers.buildInitialState(localsForBuildInitialState))
        ),
        enhancer,
        localsForBuildInitialState,
        extraThunkArguments: { fetchClient },
      });

      // TODO: namespace?
      // use the store as a global for the request
      req.store = store; // eslint-disable-line no-param-reassign
      req.clientModuleMapCache = getClientModuleMapCache();
      return next();
    } catch (err) {
      console.error('error creating store for request', err);
      return renderStaticErrorPage(res);
    }
  };
}
