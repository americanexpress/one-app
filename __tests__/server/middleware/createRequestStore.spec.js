/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import combineReducers from '@americanexpress/vitruvius/immutable';
import holocron from 'holocron';
import httpMocks from 'node-mocks-http';
import { fromJS } from 'immutable';
import { renderStaticErrorPage } from '../../../src/server/middleware/sendHtml';
import createRequestStore from '../../../src/server/middleware/createRequestStore';
import { setClientModuleMapCache } from '../../../src/server/utils/clientModuleMapCache';

jest.mock('../../../src/server/middleware/sendHtml');
jest.mock('holocron', () => {
  const actualHolocron = jest.requireActual('holocron');
  return {
    ...actualHolocron,
    createHolocronStore: jest.fn(actualHolocron.createHolocronStore),
    getModuleMap: jest.fn(),
  };
});

holocron.getModuleMap.mockImplementation(() => fromJS({
  modules: {
    'test-root': {
      baseUrl: 'https://example.com/cdn/test-root/2.2.2/',
      node: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.node.js',
        integrity: '4y45hr',
      },
      browser: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.browser.js',
        integrity: 'nggdfhr34',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js',
        integrity: '7567ee',
      },
    },
  },
}));

setClientModuleMapCache(holocron.getModuleMap().toJS());

describe('createRequestStore', () => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'error').mockImplementation(() => {});

  let req;
  let res;
  let next;
  let reducers;

  beforeAll(() => {
    global.fetch = () => Promise.resolve({ data: 'data' });
    holocron.createHolocronStore.mockClear();
  });

  afterAll(() => {
    global.fetch = undefined;
  });

  beforeEach(() => {
    jest.clearAllMocks();
    req = httpMocks.createRequest({
      headers: {
        'correlation-id': 'abc123',
      },
    });
    res = httpMocks.createResponse({ req });

    res.status = jest.fn(res.status);
    res.send = jest.fn(res.send);
    res.end = jest.fn(res.end);
    res.cookie = jest.fn(res.cookie);

    next = jest.fn();

    reducers = jest.fn(
      (...args) => combineReducers({
        appReducer: jest.fn((v) => v || true),
        config: jest.fn((v) => v || true),
      })(...args)
    );
    reducers.buildInitialState = jest.fn(() => fromJS({ appReducer: 'fizzy' }));
  });

  it('returns a function', () => {
    const middleware = createRequestStore({ reducers });
    expect(middleware).toBeInstanceOf(Function);
  });

  it('should add a store to the request object', () => {
    const middleware = createRequestStore({ reducers });
    middleware(req, res, next);
    expect(req.store).toBeTruthy();
    expect(next).toHaveBeenCalled();
  });

  it('should add the client holocron module map cache to the request object', () => {
    const middleware = createRequestStore({ reducers });
    middleware(req, res, next);
    expect(req.clientModuleMapCache).toMatchSnapshot();
    expect(next).toHaveBeenCalled();
  });

  it('should send the static error page when there is an error', () => {
    const middleware = createRequestStore({ reducers: null });
    middleware(req, res, next);
    // eslint-disable-next-line no-console
    expect(console.error).toHaveBeenCalled();
    expect(renderStaticErrorPage).toHaveBeenCalledWith(res);
  });

  describe('fetch', () => {
    it('should set the store up with fetchClient', async () => {
      const middleware = createRequestStore({ reducers });
      middleware(req, res, next);
      const {
        extraThunkArguments: { fetchClient },
      } = holocron.createHolocronStore.mock.calls[0][0];
      expect(fetchClient).toBeInstanceOf(Function);
    });
  });

  describe('useBodyForBuildingTheInitialState', () => {
    it('uses the request body as the locals for initial state when useBodyForBuildingTheInitialState is true', () => {
      const middleware = createRequestStore(
        { reducers },
        { useBodyForBuildingTheInitialState: true }
      );
      req.body = { some: 'form data', that: 'was posted' };
      middleware(req, res, next);
      expect(reducers.buildInitialState).toHaveBeenCalledTimes(1);
      expect(reducers.buildInitialState.mock.calls[0][0]).toHaveProperty('req.body.some', 'form data');
      expect(reducers.buildInitialState.mock.calls[0][0]).toHaveProperty('req.body.that', 'was posted');
    });

    it('does not use the request body as the locals for initial state when useBodyForBuildingTheInitialState is not given', () => {
      const middleware = createRequestStore({ reducers });
      req.body = { some: 'other form data', that: 'was acquired' };
      middleware(req, res, next);
      expect(reducers.buildInitialState).toHaveBeenCalledTimes(1);
      expect(reducers.buildInitialState.mock.calls[0][0]).not.toHaveProperty('req.body');
    });
  });

  it('should pass enhancedFetch into createHolocronStore', () => {
    createRequestStore({ reducers })(req, res, next);
    expect(holocron.createHolocronStore.mock.calls[0][0]).toHaveProperty('extraThunkArguments.fetchClient');
  });
});
