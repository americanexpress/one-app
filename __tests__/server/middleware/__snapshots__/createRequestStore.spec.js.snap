// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRequestStore should add the client holocron module map cache to the request object 1`] = `
Object {
  "browser": Object {
    "modules": Object {
      "test-root": Object {
        "baseUrl": "https://example.com/cdn/test-root/2.2.2/",
        "browser": Object {
          "integrity": "nggdfhr34",
          "url": "https://example.com/cdn/test-root/2.2.2/test-root.browser.js",
        },
      },
    },
  },
  "legacyBrowser": Object {
    "modules": Object {
      "test-root": Object {
        "baseUrl": "https://example.com/cdn/test-root/2.2.2/",
        "legacyBrowser": Object {
          "integrity": "7567ee",
          "url": "https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js",
        },
      },
    },
  },
}
`;
